reference/vane-apis/behn.md

```diff
@@ -0,0 +1,366 @@
++++
+title = "Behn"
+weight = 2
+template = "doc.html"
++++
+
+# Behn
+
+In this document we describe the public interface for Behn. Namely, we describe
+each `task` that Behn can be `pass`ed, and which `gift`(s) Behn can `give` in return.
+
+
+## Tasks
+
+### `%born`
+
+Each time you start your Urbit, the Arvo kernel calls the `%born` task for Behn. When
+called, Behn gets the current time from Unix and updates its list of timers
+accordingly.
+
+#### Accepts
+
+```hoon
+[~]
+```
+
+#### Returns
+
+`%born` does not return a `gift`.
+
+#### Source
+```hoon
+::  +born: urbit restarted; refresh :next-wake and store wakeup timer duct
+  ::
+  ++  born  set-unix-wake(next-wake.state ~, unix-duct.state duct)
+```
+
+
+### `%crud`
+
+`%crud` is called whenever an error involving Behn occurs, which for now means
+that Behn has no timers in its state.
+
+When called, it checks to see if the set of timers is empty. If so, it prints
+and error report containing `error`, otherwise it passes the error to `+wake`.
+
+Behn should be the first vane to be activated in a `%wake`. Occasionally this does not
+happen, but we do not yet handle this error.
+
+#### Accepts
+
+```hoon
+[tag=@tas error=tang]
+```
+Here, `tag` is a `@tas` denoting the type of error, and `error` is an error message.
+
+#### Returns
+
+If the set of timers is nonempty when Behn is `%pass`ed a `%crud` `task`, Behn
+`%give`s a `%wake` `card` to itself.
+
+```hoon
+[%wake %behn-crud-no-timer tag error]
+```
+ 
+
+#### Source
+```hoon
+++  crud
+    |=  [tag=@tas error=tang]
+    ^+  [moves state]
+    ::  behn must get activated before other vanes in a %wake
+    ::
+    ::    TODO: uncomment this case after switching %crud tags
+    ::
+    ::    We don't know how to handle other errors, so relay them to %dill
+    ::    to be printed and don't treat them as timer failures.
+    ::
+    ::  ?.  =(%wake tag)
+    ::    ~&  %behn-crud-not-first-activation^tag
+    ::    [[duct %slip %d %flog %crud tag error]~ state]
+    ::
+    ?:  =(~ timers.state)  ~|  %behn-crud-no-timer^tag^error  !!
+    ::
+    (wake `error)
+ ```
+ 
+
+### `%drip`
+
+`%drip` is utilized to delay `%give`ing a `gift`.
+
+`%drip` allows one to delay `gift`s until a given condition is met. For example,
+if Clay crashes and you do not wish for an app to then crash when it tries to
+access Clay before it is restarted, `%drip` can delay that access until Clay has
+been restored. 
+
+`%drip` only handles `gift`s, and can only schedule `gift`s for as soon as
+possible after the prescribed condition is met.
+
+#### Accepts
+
+```hoon
+mov=vase
+```
+
+`%drip` takes in a `%give` `move` in a `vase`.
+
+#### Returns
+
+In response to a `%drip` `task`, Behn will `%give` a `%meta` `gift` containing
+the `gift` originally `%give`n to Behn when `%drip` was first called.
+
+Is that really correct? It looks like the gift returned is actually a response to
+a pass it passed itself, not the original event?
```

Comment:
It is correct.  Behn becomes its own client for `%drip`.  Basically,

```
Client -- %drip --> Behn -- %wait --> Behn -- %wake --> Behn -- %give --> Target
```

Actually, the "Client" and "Target" are the same from Behn's perspective.  The client should call this with `%slip` instead of `%pass`, which lets the client impersonate the target.

Reply:
Thanks, I will include this information. Where does giving a `%meta` gift to itself show up in this sequence?

Reply:
The last `%give --> Target` is a `%meta` give, since it doesn't know what type of gift the client wants to ultimately send to the target.

Reply:
I split this up between a few commits so I'll just paste the revised drip entry here:
 

### `%drip`

`%drip` allows one to delay a `gift` until a timer set for `now` fires.

A Client gives Behn a `%drip` task wrapping a `gift` to be `give`n to a Target.
This launches a sequence of `move`s as written here:

```
Client -- %drip --> Behn -- %wait --> Behn -- %wake --> Behn -- %meta --> Target
```
Here the `%meta` move is a wrapper for the `%gift` inside of the initial `%drip` wrapper.

`%drip` only handles `gift`s, and can only schedule `gift`s for as soon as
possible after the prescribed timer fires.


#### Accepts

```hoon
mov=vase
```

`%drip` takes in a `%give` `move` in a `vase`.

#### Returns

In response to a `%drip` `task`, Behn will `%give` a `%meta` `gift` containing
the `gift` originally `%give`n to Behn when `%drip` was first called. That makes
Behn its own client for `%drip`.

#### Example


Say an app (the Target) is subscribed to updates from Clay (the Client). If Clay `%give`s
updates to the app directly and the app crashes, this may cause Clay to crash as
well. If instead Clay `%pass`es Behn a `%drip` `task` wrapping the update
`gift`, Behn will set a timer for `now` that, when fired, will cause the update
`gift` to be given. If it causes a crash then it will have been in response to
the `%drip` move, thereby isolating Clay from the crash. Thus `%drip` acts as a sort of buffer against cascading
sequences of crashes.

Reply:
- The initial client's call to Behn is not actually a `%give`, it's more like a `%pass`.  Technically it's a `%slip`, which "`%pass` but impersonate this duct", which is a hack.  I would use the word "pass", but maybe not "`%pass`"

- The `%meta` that goes to the target vane is a `[%give %meta ...]`, so I would still write it as a `%give` in the little diagram.

Reply:
Thanks, I'll poke around and try to figure out what %slip is, since we should definitely mention that it is not an ordinary %pass.

I wrote %meta instead of %give since the other labels in the sequence say which card it is instead of the type of card (type probably isn't the right word) and I thought it should be consistent. What I should actually do is make a small diagram in the style of the move trace diagram that has all of the relevant information.

Reply:
I'm still confused about which part is the `%slip`. Does the client `%slip` a `%drip` to Behn wrapping a `%gift`, or does it `%pass` Behn a `%drip` wrapping a `%slip`?

Reply:
Oh you're right, `%meta` is correct in the diagram.

`%slip` is a sibling of `%pass` and `%give` in +wind.  Technically, `%pass` says "push this wire onto the duct and transfer control to this other vane", `%give` says "pop a wire off the duct and transfer control back to whoever pushed the wire on", and `%slip` says "transfer control to this other vane but don't push or pop a wire".  This means a `%give` in response to a `%slip` will go directly back to the caller of the vane that sent the `%slip`.  This is super hacky and shouldn't really exist.

Reply:
Since this doc is the wrong place to discuss slip, I decided to just say that `%drip` uses `%slip` and made an issue for me to document `%slip` elsewhere. I started on a diagram yesterday, but didn't come up with a satisfactory way to represent `%slip`s yet so I'm just gonna stick with the text diagram for now.

How's this look?


### `%drip`

`%drip` allows one to delay a `gift` until a timer set for `now` fires.

A Client `%slip`s Behn a `%drip` task wrapping a `gift` to be `give`n to a Target.
This launches a sequence of `move`s as written here:

```
Client -- %slip %drip --> Behn -- %pass %wait --> Behn -- %give %wake --> Behn -- %give %meta --> Target
```
Here the `%meta` `move` is a wrapper for the `%gift` inside of the initial `%drip` wrapper.

`%drip` only handles `gift`s, and can only schedule `gift`s for as soon as
possible after the prescribed timer fires.


#### Accepts

```hoon
mov=vase
```

`%drip` takes in a `%give` `move` in a `vase`.

#### Returns

In response to a `%drip` `task`, Behn will `%pass` a `%wait` to itself, which
then triggers a `%wake` `gift` to itself, which then causes Behn to `%give` a `%meta` `gift` containing
the `gift` originally `%give`n to Behn when `%drip` was first called. That makes
Behn its own client for `%drip`.

#### Example


Say an app (the Target) is subscribed to updates from Clay (the Client). If Clay `%give`s
updates to the app directly and the app crashes, this may cause Clay to crash as
well. If instead Clay `%pass`es Behn a `%drip` `task` wrapping the update
`gift`, Behn will set a timer for `now` that, when fired, will cause the update
`gift` to be given. If it causes a crash then it will have been in response to
the `%drip` move, thereby isolating Clay from the crash. Thus `%drip` acts as a sort of buffer against cascading
sequences of crashes.

Reply:
Perfect!